facilitator/src/logging.rs

```diff
@@ -32,28 +42,84 @@ struct LogEntry {
     message: String,
     severity: Severity,
     time: String,
+
+    #[serde(flatten)]
+    extra: HashMap<String, Value>,
 }
 
 impl LogEntry {
-    fn from_record(fmt: &Formatter, record: &Record) -> LogEntry {
+    fn new(message: String, severity: Severity, time: String) -> Self {
         LogEntry {
-            severity: Severity::from_log_level(record.level()),
-            time: fmt.timestamp().to_string(),
-            message: record.args().to_string(),
+            message,
+            severity,
+            time,
+            extra: HashMap::new(),
         }
     }
 }
 
+impl convert::From<&Record<'_>> for LogEntry {
+    fn from(record: &Record) -> Self {
+        let severity: Severity = record.level().into();
+        let message: String = record.msg().to_string();
+        let time = chrono::Utc::now().to_rfc3339();
+
+        let mut entry = LogEntry::new(message, severity, time);
+        let _ = record.kv().serialize(record, &mut entry);
+
+        entry
+    }
+}
+
+/// slog::Serializer allows us to handler the KV records emitted by slog.
+impl slog::Serializer for LogEntry {
+    fn emit_arguments(&mut self, k: slog::Key, v: &std::fmt::Arguments) -> slog::Result {
+        let value = match serde_json::to_value(v) {
+            Ok(value) => value,
+            Err(_) => Value::String(v.to_string()),
+        };
+
+        self.extra.insert(k.to_string(), value);
+
+        Ok(())
+    }
+}
+
+/// JSONDrain is a custom slog drain implementation that implements the
+/// slog::Drain trait. It acts as a final output for the log messages.
+#[derive(Debug)]
+struct JSONDrain {}
```

Comment:
Would [slog-json](https://docs.rs/slog-json/2.3.0/slog_json/) save us some code? Seems like a `slog-json::Json` hooked up to [`std::io::stdout`](https://doc.rust-lang.org/std/io/fn.stdout.html) would do the trick.